syntax = "proto3";

package sentinel.v1;

import "google/protobuf/timestamp.proto";

// The core management service for the planetary analysis mesh.
service OrchestratorService {
  rpc RegisterNode(RegisterNodeRequest) returns (RegisterNodeResponse);
  rpc SubmitScan(SubmitScanRequest) returns (SubmitScanResponse);
  rpc StreamTelemetry(StreamTelemetryRequest) returns (stream TelemetryUpdate);
}

service ConsensusService {
  rpc RequestVote(VoteRequest) returns (VoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}

message VoteRequest {
  uint64 term = 1;
  string candidate_id = 2;
  uint64 last_log_index = 3;
  uint64 last_log_term = 4;
}

message VoteResponse {
  uint64 term = 1;
  bool vote_granted = 2;
}

message AppendEntriesRequest {
  uint64 term = 1;
  string leader_id = 2;
  uint64 prev_log_index = 3;
  uint64 prev_log_term = 4;
  repeated RaftLogEntry entries = 5;
  uint64 leader_commit = 6;
}

message RaftLogEntry {
  uint64 term = 1;
  uint64 index = 2;
  string command = 3;
}

message AppendEntriesResponse {
  uint64 term = 1;
  bool success = 2;
}

message RegisterNodeRequest {
  string node_id = 1;
  string region = 2;
  string capabilities = 3; // JSON or bitmask
  bytes public_key_pqc = 4;
}

message RegisterNodeResponse {
  bool success = 1;
  string assignment_token = 2;
}

message SubmitScanRequest {
  string repo_url = 1;
  string tenant_id = 2;
  repeated string analysis_phases = 3; // ["taint", "formal", "symbolic"]
}

message SubmitScanResponse {
  string job_id = 1;
  google.protobuf.Timestamp estimated_completion = 2;
}

message StreamTelemetryRequest {
  string job_id = 1;
}

message TelemetryUpdate {
  string job_id = 1;
  float progress = 2;
  string current_file = 3;
  uint32 vulnerabilities_found = 4;
}

// Inference Service for AI-driven code analysis
service InferenceService {
  rpc InferLogic(InferenceRequest) returns (InferenceResponse);
  rpc AnalyzeVulnerability(AnalysisRequest) returns (AnalysisResponse);
}

message InferenceRequest {
  string subgraph_id = 1;
  repeated uint64 nodes = 2;
  repeated FeatureVector features = 3;
  bytes pqc_signature = 4;
}

message FeatureVector {
  repeated float values = 1;
}

message InferenceResponse {
  repeated float embedding = 1;
  float confidence = 2;
  bool logic_clone_detected = 3;
}

message AnalysisRequest {
  string file_path = 1;
  string code_snippet = 2;
  string vulnerability_type = 3;
}

message AnalysisResponse {
  string explanation = 1;
  string proposed_fix = 2;
  float confidence = 3;
}
